src/sparseml/modifiers/smoothquant/pytorch.py

```diff
@@ -0,0 +1,217 @@
+
+                hidden_dim = out.shape[-1]
+                out = out.view(-1, hidden_dim).abs()
+                latest_mins = torch.min(out, dim=0)[0]
+                latest_maxes = torch.max(out, dim=0)[0]
+
+                if layer_name in self.scales_:
+                    self.scales_[layer_name].min_channel_vals = torch.minimum(
+                        self.scales_[layer_name].min_channel_vals, latest_mins
+                    )
+                    self.scales_[layer_name].max_channel_vals = torch.maximum(
+                        self.scales_[layer_name].max_channel_vals, latest_maxes
+                    )
+                else:
+                    self.scales_[layer_name] = SmoothQuantScale(
+                        min_channel_vals=latest_mins, max_channel_vals=latest_maxes
+                    )
+
+            return hook_fn
+
+        for mapping in self.resolved_mappings_:
+            name = mapping.smooth_name
+            layer = mapping.smooth_layer
+            self.hooks_.append(layer.register_forward_hook(create_hook_fn(name)))
+
+    @torch.no_grad()
+    def _calibrate(self, model: ModifiableModelPyTorch, calibration_dataloader: List):
+        """
+        Catch the output dynamic ranges of each layer that will be smoothed by running
+        forward passes with calibration_dataloader
+        """
+        _LOGGER.info("Running SmoothQuant scale calibration...")
+        if not calibration_dataloader:
+            raise ValueError(
+                "Calibration data loader not set, must populate the calib_data field of"
+                " SparseSession to run the SmoothQuant modifier"
+            )
+
+        model.model.eval()
+
+        run_calibration_forward(
+            model.model,
+            calibration_dataloader,
+            self.num_calibration_steps,
+            self.calibration_function,
+        )
+
+        # remove the hooks now that we are done calibrating
+        for hook in self.hooks_:
+            hook.remove()
+        del self.hooks_
+
+    @torch.no_grad()
+    def _apply_smoothing(self):
+        """
+        After calibration, apply smoothing to the activations and push the transform
+        into the following weights by applying the inverse to each balance weight.
+
+        Y = (Xdiag(scales)^(-1) * diag(scales)W) where W is the to_balance weights and
+        X is the to_smooth weights
+
+        This modifies the weights of the model in-place.
+        """
+        _LOGGER.info("Smoothing activation scales...")
+        for mapping in self.resolved_mappings_:
+            activation_scales = (  # get dynamic range for each activation channel
+                self.scales_[mapping.smooth_name].max_channel_vals
+                - self.scales_[mapping.smooth_name].min_channel_vals
+            )
+            smooth_layer = mapping.smooth_layer
+            balance_layers = mapping.balance_layers
+
+            # get the channel-wise dynamic range for each layer to be balanced
```

Comment:
This section between lines 192 and 205 is the computation of smoothing scales I referred to in the comment above. I suggest moving this to a separate method so we can override it when implementing LogarithmicEqualizationModifier.